nixos/modules/hardware/device-tree.nix

```
@@ -4,7 +4,106 @@ with lib;
 
 let
   cfg = config.hardware.deviceTree;
+
+  overlayType = types.submodule {
+    options = {
+      name = mkOption {
+        type = types.str;
+        description = ''
+          Name of this overlay
+        '';
+      };
+
+      dtsFile = mkOption {
+        type = types.nullOr types.path;
+        description = ''
+          Path to .dts overlay file, overlay is applied to
+          each .dtb file matching "compatible" of the overlay.
+        '';
+        default = null;
+        example = literalExample "./dts/overlays.dts";
+      };
+
+      dtsText = mkOption {
+        type = types.nullOr types.str;
+        default = null;
+        description = ''
+          Literal DTS contents, overlay is applied to
+          each .dtb file matching "compatible" of the overlay.
+        '';
+        example = literalExample ''
+          /dts-v1/;
+          /plugin/;
+          / {
+                  compatible = "raspberrypi";
+                  fragment@0 {
+                          target-path = "/soc";
+                          __overlay__ {
+                                  pps {
+                                          compatible = "pps-gpio";
+                                          status = "okay";
+                                  };
+                          };
+                  };
+          };
+        '';
+      };
+
+      dtboFile = mkOption {
+        type = types.nullOr types.path;
+        default = null;
+        description = ''
+          Path to .dtbo compiled overlay file.
+        '';
+      };
+    };
+  };
+
+  # this requires kernel package
+  dtbsWithSymbols = pkgs.stdenv.mkDerivation {
+        name = "dtbs-with-symbols";
+        inherit (cfg.kernelPackage) src nativeBuildInputs;
+        patches = map (patch: patch.patch) cfg.kernelPackage.kernelPatches;
+
+        buildPhase = ''
+          substituteInPlace scripts/Makefile.lib \
+            --replace 'DTC_FLAGS += $(DTC_FLAGS_$(basetarget))' 'DTC_FLAGS += $(DTC_FLAGS_$(basetarget)) -@'
+          make ${pkgs.stdenv.hostPlatform.platform.kernelBaseConfig}
+          make dtbs
+        '';
+
+        installPhase = ''
+          make dtbs_install INSTALL_DTBS_PATH=$out/dtbs
+        '';
+      };
+
+  filterDTBs = src: if isNull cfg.filter
+    then "${src}/dtbs"
+    else
+      pkgs.runCommand "dtbs-filtered" {} ''
+        mkdir -p $out
+        for f in $( find ${src} -maxdepth 2 -type f -name '${cfg.filter}' ); do
+          cp -v $f $out/
+        done
+      '';
+
+  compileDTS = name: f: pkgs.runCommand "${name}-dtbo" {} ''
+    ${pkgs.dtc}/bin/dtc -I dts ${f} -O dtb -@ -o $out
+  '';
```

For cross-compiling I had to make this change or else the ARM version  of `dtc` were used.

```suggestion
  compileDTS = name: f: pkgs.callPackage({dtc}: pkgs.runCommand "${name}-dtbo" {
    nativeBuildInputs = [dtc];
  } ''
    dtc -I dts ${f} -O dtb -@ -o $out
  '') {};
```

---

The principle here is OK too, I think if we're passing `nativeBuildInputs` we may as well use `mkDerivation` here with `buildCommand`.

---

Fixed using `mkDerivation`.